File: Leet-Code/rotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 #include <deque>
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == nullptr)
            return nullptr;

        deque<ListNode*> queue;
        ListNode* current = head;

        while(current != NULL)
        {
            queue.push_back(current);
            current = current->next;
        }
        if(queue.size() == 1)
            return head;

        else
        {
            int rotations = k % queue.size();
            for(int rotation = 0; rotation < rotations; ++rotation)
            {
                current = queue.back();
                queue.pop_back();
                queue.back()->next = NULL;

                current->next = queue.front();
                queue.push_front(current);
            }
            return queue.front();
        }
    }
};

File: Leet-Code/rotateList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next 

from collections import deque

class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None:
            return head

        current = head
        dq = deque()
        while current is not None:
            dq.append(current)
            current = current.next
   
        if len(dq) is 1:
            return head
        else:
            rotations = k % len(dq)
            for rotation in range(0, rotations):
                current = dq.pop()
                current.next = dq[0]
                dq.appendleft(current)
                dq[-1].next=None

        return dq[0]
                        